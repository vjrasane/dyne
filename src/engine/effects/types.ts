import { Dispatch } from "../core/dispatch";

export type PromiseableF<A> = () => Promiseable<A>;

export type Promiseable<A> = Promise<A> | A;

export type CmdFx<A> = PromiseableF<A> | Promiseable<A>;

export type ProcessorF<I, O> = (res: I) => O;

export type Processor<I, O> = ProcessorF<I, O> | ((res: I) => I) | O;

export type Failure<O> = Processor<Error, O>;

export type Subscriber = (dispatch: Dispatch) => void;

export type Subscribe<A> = (dispatch: Dispatch) => A;

export type Canceller = () => void;

export type Cancel<A> = (sub: A) => void;
