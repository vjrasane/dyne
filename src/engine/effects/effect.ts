import { Dispatch } from "../core/dispatch";

export default class Effect<A> {
  effect: A;

  constructor(effect: A) {
    this.effect = effect;
  }

  execute = (_: Dispatch): void => {};
}
