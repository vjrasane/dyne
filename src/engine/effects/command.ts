import { utils as U } from "dyne-commons";
import { CmdFx, ProcessorF, PromiseableF, Processor, Failure } from "./types";
import Effect from "./effect";
import { Effectable } from "./return";
import { Dispatch } from "../core/dispatch";

export class Command<I, O> extends Effect<CmdFx<I>> {
  success: Processor<I, O>;
  failure: Failure<O>;

  constructor(cmd: CmdFx<I>, success?: Processor<I, O>, failure?: Failure<O>) {
    super(cmd);
    this.success = U.exists(success) ? success : U.identity;
    this.failure = U.exists(failure) ? failure : U.identity;
  }

  execute = (dispatch: Dispatch): Promise<void> =>
    processCommand(dispatch, this);
}

const processResult = (dispatch: Dispatch) => <I, O>(
  processor: Processor<I, O>,
  result: I
): void =>
  dispatch(
    U.isFunction(processor) ? (<ProcessorF<I, O>>processor)(result) : processor
  );

const processCommand = async <I, O>(
  dispatch: Dispatch,
  cmd: Command<I, O>
): Promise<void> => {
  const process: <I, O>(
    processor: Processor<I, O>,
    result: I
  ) => void = processResult(dispatch);

  try {
    const result = U.isFunction(cmd.effect)
      ? (<PromiseableF<I>>cmd.effect)()
      : cmd.effect;
    process(cmd.success, await result);
  } catch (error) {
    process(cmd.failure, error);
  }
};

export function Cmd<I, O>(
  cmd: CmdFx<I>,
  success?: Processor<I, O>,
  failure?: Failure<O>
): Command<I, O> {
  return new Command<I, O>(cmd, success, failure);
}

Cmd.wrap = <A>(wrappable: Effectable<A>): Effect<CmdFx<A>> => {
  if (wrappable instanceof Effect) {
    return wrappable;
  } else {
    return Cmd(wrappable);
  }
};
