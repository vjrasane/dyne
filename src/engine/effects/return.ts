import Effect from "./effect";
import { Cmd } from "./command";
import { CmdFx } from "./types";

export type Effectable<A> = Effect<A> | CmdFx<A>;

export default class Return<A> {
  model: A;
  effects: Effectable<any>[];

  constructor(model: A, ...effects: Effectable<any>[]) {
    this.model = model;
    this.effects = effects.map(Cmd.wrap);
  }

  map = <B>(mapper: (model: A) => B): Return<B> =>
    new Return(mapper(this.model), ...this.effects);

  bind = <B>(mapper: (model: A) => Return<B>): Return<B> => {
    const ret = mapper(this.model);
    return new Return(ret.model, ...this.effects, ...ret.effects);
  };
}
