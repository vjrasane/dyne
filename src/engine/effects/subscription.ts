import Effect from "./effect";
import { types as T } from "dyne-commons";
import { Cmd, Command } from "./command";
import { Dispatch, msgEventListener } from "../core/dispatch";
import { Subscriber, Canceller, Subscribe, Cancel } from "./types";

export class Subscription extends Effect<Subscriber> {
  canceller: Canceller;

  constructor(subscriber: Subscriber, canceller: Canceller) {
    super(subscriber);
    this.canceller = canceller;
  }

  execute = (dispatch: Dispatch): void => this.effect(dispatch);
  cancel = (): Command<void, void> => Cmd(this.canceller);
}

export const Sub = <A>(subscribe: Subscribe<A>, cancel: Cancel<A>) => {
  let sub: A;
  const subscriber: Subscriber = (dispatch: Dispatch): void => {
    sub = subscribe(dispatch);
  };
  const canceller: Canceller = (): void => cancel(sub);
  return new Subscription(subscriber, canceller);
};

Sub.interval = <A>(effect: Effect<A> | A, duration: number): Subscription =>
  Sub(
    (dispatch: Dispatch): NodeJS.Timer =>
      setInterval((): void => dispatch(effect), duration),
    (interval: NodeJS.Timer): void => clearInterval(interval)
  );

Sub.listen = (trigger: string, effect: T.Generator<object>): Subscription =>
  Sub(
    (dispatch: Dispatch): T.Sink => {
      const listener: T.Sink = msgEventListener(effect, dispatch);
      window.addEventListener(trigger, listener);
      return listener;
    },
    (listener: T.Sink): void => window.removeEventListener(trigger, listener)
  );
