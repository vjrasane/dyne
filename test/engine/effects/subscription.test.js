import { utils } from "dyne-commons";
import { Sub } from "../../../src/engine/effects/subscription";
import { Cmd } from "../../../src/engine/effects/command";

const { identity } = utils;

jest.useFakeTimers();

describe("subscription", () => {
  let dispatch;
  beforeEach(() => {
    dispatch = jest.fn();
  });

  it("sets interval with simple effect", () => {
    Sub.interval("effect", 100).execute(dispatch);

    jest.advanceTimersByTime(1000);
    expect(dispatch).toHaveBeenCalledWith("effect");
    expect(dispatch).toHaveBeenCalledTimes(10);
  });

  it("sets interval with command effect", () => {
    const cmd = Cmd(() => "effect");
    Sub.interval(cmd, 100).execute(dispatch);

    jest.advanceTimersByTime(1000);
    expect(dispatch).toHaveBeenCalledWith(cmd);
    expect(dispatch).toHaveBeenCalledTimes(10);
  });

  it("cancels interval", async () => {
    const sub = Sub.interval("effect", 100);
    sub.execute(dispatch);
    await sub.cancel().execute(dispatch);
    dispatch.mockClear();

    jest.advanceTimersByTime(1000);
    expect(dispatch).not.toHaveBeenCalled();
  });

  it("listens to window events", () => {
    Sub.listen("click", event => ({ msg: "click", event })).execute(dispatch);
    window.dispatchEvent(new window.Event("click"));
    expect(dispatch.mock.calls).toMatchSnapshot();
  });

  it("stops listening to window events", async () => {
    const sub = Sub.listen("click", event => ({ msg: "click", event }));
    await sub.cancel().execute(dispatch);
    dispatch.mockClear();

    window.dispatchEvent(new window.Event("click"));
    expect(dispatch).not.toHaveBeenCalled();
  });
});
