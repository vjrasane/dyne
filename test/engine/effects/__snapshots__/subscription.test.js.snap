// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`subscription listens to window events 1`] = `
Array [
  Array [
    Object {
      "event": Event {
        "isTrusted": false,
      },
      "msg": "click",
    },
  ],
]
`;
